### ISSUE TITLE
Add option to sort `__all__` according to PEP8

### ISSUE THREAD

**Issue by nstarman** [astral-sh/ruff#9324]
In https://pep8.org/#module-level-dunder-names it says that `__all__` "should be placed after the module docstring but before any import statements except from `__future__` imports".
It would be great to have an isort-style rule that moves `__all__` to before the imports.
There are definitely cases when `__all__` shouldn't be moved, e.g. when it is dynamically defined, but if `__all__` contains only strings then it should be safe to move.

**Comment by Skylion007** [comment]
For reference, isort already has this an option so we might just be able to add it to the isort rules: https://github.com/PyCQA/isort/commit/44e3b5d179da5033ce23f796b014e74f3a3259cd

**Comment by charliermarsh** [comment]
Does that open sort the _members_ of `__all__`, or move the `__all__` itself? I couldn't tell from the PR.

**Comment by alchzh** [comment]
> For reference, isort already has this an option so we might just be able to add it to the isort rules: [PyCQA/isort@44e3b5d](https://github.com/PyCQA/isort/commit/44e3b5d179da5033ce23f796b014e74f3a3259cd)

That is a different rule that is tracked by #1198. The title of this issue is a bit confusing. `--srx` sorts the strings inside of `__all__`, while this request is about moving the `__all__` (and `__version__`, `__author__`, etc.) _assignment statements_ to the top of the file, e.g.

Bad:
```py
"""Dummy module"""

from __future__ import annotations

import os

__all__ = ["foo"]
```

Good:
```py
"""Dummy module"""

from __future__ import annotations

__all__ = ["foo"]

import os
```

@nstarman , I think the title of this issue should be changed to something like "Add option to move module level dunders to top of file"

**Comment by alchzh** [comment]
It's worth noting that before 2016 the PEP8 mandated the opposite,
> Put any relevant ``__all__`` specification after the imports.

This was changed to the current text in
https://bugs.python.org/issue27187 ([patch](https://bugs.python.org/file43132/issue-27187-patch1.txt))
https://github.com/python/peps/commit/0aa70aebf102d352b8476f04509a369bf3db276c

but the style checkers only relaxed the previous rule instead of implementing the new one https://github.com/PyCQA/pycodestyle/pull/523 https://github.com/PyCQA/pycodestyle/issues/615 , so there's been no pressure to change it in existing codebases. It seems like most Python code follows the older convention still, since people tend to go by whatever their style checker / formatter says instead of reading PEP8 themselves.

**Comment by btakita** [comment]
It seems that `__all__` at the top of the file is well supported now. It reduces the line count in the file due to reduction of surrounding empty lines. It is easier to find `__all__`...**the public api**...when there are many imports in the module.

Pycharm also nudges to have `__all__` at the top of the file, as auto-import will add the import below `__all__` by default.

It would be great to have this as a linting option...particularly for changing a codebase to be PEP8 compliant.

> It's worth noting that before 2016 the PEP8 mandated the opposite,

2016 was a decade ago...
